### PR TITLE
Introduce UseAzureSynapse and UseAzureSql

### LINKED ISSUE #33816: Introduce UseAzureSql/UseAzureSynapse as alternatives to UseSqlServer
We're increasingly seeing T-SQL divergence, where certain features are not available e.g. on Azure Synapse (e.g. #33555), and some features are only available Azure SQL (e.g. the new JSON type, the new vector search capabilities). To account for this divergence, we'll allow users to specify which database is being targeted by providing UseAzureSql/UseAzureSynapse alternatives to UseSqlServer.

Note: we'd need to duplicate all current variants of UseSqlServer.

### PR DESCRIPTION
**Pull request by cincuranet**
Fixes #33816

### CODE REVIEW

File: src/EFCore.SqlServer/Extensions/AzureSqlDbContextOptionsBuilderExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+
+/// <summary>
+///     Azure SQL specific extension methods for <see cref="DbContextOptionsBuilder" />.
+/// </summary>
+/// <remarks>
+///     See <see href="https://aka.ms/efcore-docs-dbcontext-options">Using DbContextOptions</see>, and
+///     <see href="https://aka.ms/efcore-docs-sqlserver">Accessing Azure SQL databases with EF Core</see>
+///     for more information and examples.
+/// </remarks>
+public static class AzureSqlDbContextOptionsExtensions
```
Comment (AndriySvyryd) [inline src/EFCore.SqlServer/Extensions/AzureSqlDbContextOptionsBuilderExtensions.cs:17]: Move to SqlServerDbContextOptionsExtensions

File: src/EFCore.SqlServer/Extensions/AzureSynapseDbContextOptionsBuilderExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+
+/// <summary>
+///     Azure Synapse specific extension methods for <see cref="DbContextOptionsBuilder" />.
+/// </summary>
+/// <remarks>
+///     See <see href="https://aka.ms/efcore-docs-dbcontext-options">Using DbContextOptions</see>, and
+///     <see href="https://aka.ms/efcore-docs-sqlserver">Accessing Azure Synapse databases with EF Core</see>
+///     for more information and examples.
+/// </remarks>
+public static class AzureSynapseDbContextOptionsExtensions
```
Comment (AndriySvyryd) [inline src/EFCore.SqlServer/Extensions/AzureSynapseDbContextOptionsBuilderExtensions.cs:17]: Move to SqlServerDbContextOptionsExtensions

File: test/EFCore.SqlServer.FunctionalTests/SqlServerConfigPatternsTest.cs [hunk 3]
```diff
@@ -480,17 +482,20 @@ private class NorthwindContext(bool azure) : DbContext
...
+                        .UseAzureSql(SqlServerNorthwindTestStoreFactory.NorthwindConnectionString);
+                }
+                else
+                {
+                    optionsBuilder
+                        .UseSqlServer(SqlServerNorthwindTestStoreFactory.NorthwindConnectionString);
+                }
+            }
 
             protected override void OnModelCreating(ModelBuilder modelBuilder)
```
Comment (AndriySvyryd) [inline test/EFCore.SqlServer.FunctionalTests/SqlServerConfigPatternsTest.cs:500]: Add a test where `UseAzureSql`, `UseSqlServer` with `EnableRetryOnFailure` and `UseAzureSynapse` are called in different orders. Validate that the execution strategy is not clobbered.

File: src/EFCore.SqlServer/Extensions/AzureSqlServiceCollectionExtensions .cs [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    ///     This is not recommend other than for some advanced scenarios.
+    /// </remarks>
+    /// <param name="serviceCollection">The <see cref="IServiceCollection" /> to add services to.</param>
+    /// <returns>
+    ///     The same service collection so that multiple calls can be chained.
+    /// </returns>
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public static IServiceCollection AddEntityFrameworkAzureSql(this IServiceCollection serviceCollection)
+    {
+        new EntityFrameworkRelationalServicesBuilder(serviceCollection)
```
Comment (AndriySvyryd) [inline src/EFCore.SqlServer/Extensions/AzureSqlServiceCollectionExtensions .cs:95]: Extract common code between this and the other methods

File: src/EFCore.SqlServer/Query/Internal/SqlServerQueryableMethodTranslatingExpressionVisitor.cs [hunk 5]
```diff
@@ -228,9 +226,9 @@ protected override Expression VisitExtension(Expression extensionExpression)
         IProperty? property,
         string tableAlias)
     {
-        if (_sqlServerCompatibilityLevel < 130)
+        if (_sqlEngineSingletonOptions is ISqlServerSingletonOptions { CompatibilityLevel: < 130 } sqlServerSingletonOptions)
```
Comment (AndriySvyryd) [inline src/EFCore.SqlServer/Query/Internal/SqlServerQueryableMethodTranslatingExpressionVisitor.cs:229]: Why `IAzureSqlSingletonOptions` isn't checked here?

Comment (cincuranet) [reply src/EFCore.SqlServer/Query/Internal/SqlServerQueryableMethodTranslatingExpressionVisitor.cs:229]: I missed that. Thanks.

**Comment by AndriySvyryd** [comment]
@cincuranet Let me know one this is ready for review again

**Comment by cincuranet** [comment]
@AndriySvyryd This should be now ready for review.

File: src/EFCore.SqlServer/Infrastructure/Internal/SqlServerSingletonOptions.cs [hunk 6]
```diff
@@ -17,15 +17,55 @@ public class SqlServerSingletonOptions : ISqlServerSingletonOptions
...
+    /// </summary>
+    public virtual int AzureSynapseCompatibilityLevel { get; private set; } = SqlServerOptionsExtension.AzureSynapseDefaultCompatibilityLevel;
+
+    /// <summary>
+    ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+    ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+    ///     any release. You should only use it directly in your code with extreme caution and knowing that
+    ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+    /// </summary>
+    public virtual int? AzureSynapseCompatibilityLevelWithoutDefault { get; private set; }
```
Comment (AndriySvyryd) [inline src/EFCore.SqlServer/Infrastructure/Internal/SqlServerSingletonOptions.cs:68]: Are the `*CompatibilityLevelWithoutDefault` used anywhere? If not, remove them

File: src/EFCore.SqlServer/Infrastructure/Internal/SqlServerSingletonOptions.cs [hunk 7]
```diff
@@ -51,15 +96,20 @@ public virtual void Initialize(IDbContextOptions options)
...
+                || SqlServerCompatibilityLevel != sqlServerOptions.SqlServerCompatibilityLevel
+                || AzureSqlCompatibilityLevelWithoutDefault != sqlServerOptions.AzureSqlCompatibilityLevelWithoutDefault
+                || AzureSqlCompatibilityLevel != sqlServerOptions.AzureSqlCompatibilityLevel
+                || AzureSynapseCompatibilityLevelWithoutDefault != sqlServerOptions.AzureSynapseCompatibilityLevelWithoutDefault
+                || AzureSynapseCompatibilityLevel != sqlServerOptions.AzureSynapseCompatibilityLevel))
         {
             throw new InvalidOperationException(
                 CoreStrings.SingletonOptionChanged(
-                    nameof(SqlServerDbContextOptionsExtensions.UseSqlServer),
+                    $"{nameof(SqlServerDbContextOptionsExtensions.UseSqlServer)}/{nameof(SqlServerDbContextOptionsExtensions.UseAzureSql)}/{nameof(SqlServerDbContextOptionsExtensions.UseAzureSynapse)}",
```
Comment (AndriySvyryd) [inline src/EFCore.SqlServer/Infrastructure/Internal/SqlServerSingletonOptions.cs:112]: Separate into different blocks

File: src/EFCore.SqlServer/Query/Internal/SqlServerSqlExpressionFactory.cs [hunk 8]
```diff
@@ -16,7 +16,7 @@ namespace Microsoft.EntityFrameworkCore.SqlServer.Query.Internal;
 public class SqlServerSqlExpressionFactory : SqlExpressionFactory
 {
     private readonly IRelationalTypeMappingSource _typeMappingSource;
-    private readonly int _sqlServerCompatibilityLevel;
+    private readonly ISqlServerSingletonOptions _sqlServerSingletonOptions;
```
Comment (AndriySvyryd) [inline src/EFCore.SqlServer/Query/Internal/SqlServerSqlExpressionFactory.cs:19]: This appears unused, remove

File: src/EFCore.SqlServer/Infrastructure/Internal/SqlServerOptionsExtension.cs [hunk 9]
```diff
@@ -168,12 +318,13 @@ public ExtensionInfo(IDbContextOptionsExtension extension)
...
             => (SqlServerOptionsExtension)base.Extension;
 
-        public override bool IsDatabaseProvider
-            => true;
-
         public override bool ShouldUseSameServiceProvider(DbContextOptionsExtensionInfo other)
             => other is ExtensionInfo otherInfo
-                && Extension.CompatibilityLevel == otherInfo.Extension.CompatibilityLevel;
+                && Extension.EngineType == otherInfo.Extension.EngineType
+                && Extension.LegacyAzureSql == otherInfo.Extension.LegacyAzureSql
```
Comment (AndriySvyryd) [inline src/EFCore.SqlServer/Infrastructure/Internal/SqlServerOptionsExtension.cs:324]: Remove this condition `&& Extension.LegacyAzureSql == otherInfo.Extension.LegacyAzureSql`
It does not impact singleton services, so same service provider can be used even if it changes

File: src/EFCore.SqlServer/Infrastructure/Internal/SqlServerOptionsExtension.cs [hunk 10]
```diff
@@ -185,11 +336,36 @@ public override string LogFragment
...
-                    if (Extension._compatibilityLevel is int compatibilityLevel)
+                    builder
+                        .Append("EngineType=")
+                        .Append(Extension._engineType)
+                        .Append(' ');
+
+                    builder
+                        .Append("LegacyAzureSql=")
+                        .Append(Extension._legacyAzureSql)
+                        .Append(' ');
```
Comment (AndriySvyryd) [inline src/EFCore.SqlServer/Infrastructure/Internal/SqlServerOptionsExtension.cs:347]: Only add it when `true`

**Review by AndriySvyryd — APPROVED** [review]

File: src/EFCore.SqlServer/Infrastructure/Internal/SqlServerSingletonOptions.cs (hunk 7 above)
Comment (cincuranet) [reply src/EFCore.SqlServer/Infrastructure/Internal/SqlServerSingletonOptions.cs:82]: The whole `if` statement?

Comment (AndriySvyryd) [reply src/EFCore.SqlServer/Infrastructure/Internal/SqlServerSingletonOptions.cs:82]: Yes, to avoid listing all methods in a single message

Comment (cincuranet) [reply src/EFCore.SqlServer/Infrastructure/Internal/SqlServerSingletonOptions.cs]: Is [this](https://github.com/dotnet/efcore/pull/34052/commits/f73141ad84e04199b2a8e43156ed181214cddb48) fine?

Comment (AndriySvyryd) [reply src/EFCore.SqlServer/Infrastructure/Internal/SqlServerSingletonOptions.cs]: Yes, that's exactly what I had in mind

**Comment by AndriySvyryd** [comment]
Looking at this again I find it confusing that you can call `.UseSqlServer(b => UseAzureSqlCompatibilityLevel())`
Each of the `Use*` methods should accept a different builder type, then they each could just have a `UseCompatibilityLevel` method that corresponds to the target engine.
Additionally, for `ConfigureSqlEngine` the builder should have a non-obsolete method that enables a retrying execution strategy if one wasn't configured (I don't remember what we decided to name this)

**Pull request merged**